### PR TITLE
ARM64 - Optimize `i % 2`

### PR DESCRIPTION
**Pull request by TIHan**
**Description**

Added a special optimization for `i % 2` specifically when the divisor is a constant `2` as there is an even more optimal form:
```asm
and     w1, w0, #1
cmp     w0, #0
cneg    w0, w1, lt
```

example diff:
```diff
             and     w0, w27, #1
-            negs    w1, w27
-            and     w1, w1, #1
-            csneg   w0, w0, w1, mi
+            cmp     w27, #0
+            cneg    w0, w0, lt
```

**Acceptance Criteria**
- [x] Merge https://github.com/dotnet/runtime/pull/68885

[Diffs](https://dev.azure.com/dnceng/public/_build/results?buildId=1829524&view=ms.vss-build-web.run-extensions-tab)

### CODE REVIEW

**Comment by TIHan** [comment]
@dotnet/jit-contrib Is ready

**Comment by TIHan** [comment]
@kunalspathak PTAL - I will give performance metrics on this soon.

File: src/coreclr/jit/lowerarmarch.cpp [hunk 1]
```diff
@@ -739,30 +731,62 @@ void Lowering::LowerModPow2(GenTree* node)
...
-    BlockRange().InsertAfter(cns, trueExpr);
-    LowerNode(trueExpr);
+    if (divisorCnsValue == 2)
+    {
+        // {expr} % 2
 
-    GenTree* const neg = comp->gtNewOperNode(GT_NEG, type, dividend2);
-    neg->gtFlags |= GTF_SET_FLAGS;
-    BlockRange().InsertAfter(trueExpr, neg);
+        GenTreeIntCon* cns = comp->gtNewIconNode(divisorCnsValueMinusOne, type);
```
Comment (kunalspathak) [inline src/coreclr/jit/lowerarmarch.cpp:738]: Some of the things like below can be common?

```
        GenTreeIntCon* cns = comp->gtNewIconNode(divisorCnsValueMinusOne, type);
        BlockRange().InsertAfter(dividend2, cns);

        GenTree* const trueExpr = comp->gtNewOperNode(GT_AND, type, dividend, cns);
        BlockRange().InsertAfter(cns, trueExpr);
        LowerNode(trueExpr);
```

File: src/coreclr/jit/lowerarmarch.cpp [hunk 2]
```diff
@@ -739,30 +731,62 @@ void Lowering::LowerModPow2(GenTree* node)
...
 
-    GenTreeIntCon* cns = comp->gtNewIconNode(cnsValue, type);
-    BlockRange().InsertAfter(dividend2, cns);
-
-    GenTree* const trueExpr = comp->gtNewOperNode(GT_AND, type, dividend, cns);
-    BlockRange().InsertAfter(cns, trueExpr);
-    LowerNode(trueExpr);
+    if (divisorCnsValue == 2)
+    {
+        // {expr} % 2
```
Comment (kunalspathak) [inline src/coreclr/jit/lowerarmarch.cpp:736]: Similar comments like you have in `else`?

**Review by kunalspathak — COMMENTED** [review]
I would just wait to see superpmi diffs.

File: src/coreclr/jit/lowerarmarch.cpp (hunk 2 above)
Comment (TIHan) [reply src/coreclr/jit/lowerarmarch.cpp:736]: Sure

File: src/coreclr/jit/lowerarmarch.cpp (hunk 1 above)
Comment (TIHan) [reply src/coreclr/jit/lowerarmarch.cpp:738]: Yea, looks like they can be lifted up

**Comment by TIHan** [comment]
[Diffs](https://dev.azure.com/dnceng/public/_build/results?buildId=1829524&view=ms.vss-build-web.run-extensions-tab) look good, all improvements, no regressions.

**Review by kunalspathak — APPROVED** [review]
LGTM

**Pull request merged**